crypto/src/merkle.rs

```diff
@@ -1,242 +1,256 @@
 //! Merkle tree implementation.
 
 #[cfg(not(feature = "std"))]
-use alloc::{boxed::Box, format, string::String, vec, vec::Vec};
+use alloc::{format, string::String, vec::Vec};
 #[cfg(feature = "std")]
 use std::collections::VecDeque;
 
 use iroha_schema::prelude::*;
 
 use crate::HashOf;
 
-/// [Merkle Tree](https://en.wikipedia.org/wiki/Merkle_tree) used to validate and prove data at
-/// each block height.
-/// Our implementation uses binary hash tree.
+/// [Merkle Tree](https://en.wikipedia.org/wiki/Merkle_tree) used to validate `T`
 #[derive(Debug)]
-pub struct MerkleTree<T> {
-    root_node: Node<T>,
+pub struct MerkleTree<T>(Vec<Option<HashOf<T>>>);
```

Comment:
if `Hash` is made `[NonNullU8, SIZE]` maybe compiler would be able to use niche optimization to represent `None` here. I'm saying this because that would shave of byte for every node of the merkle tree

Reply:
This is something that can be considered for future work, not for this PR

Reply:
The problem with making it `NonZeroU8` is that then all bytes have to be non-zero. The odds of such a key not occurring are astronomically low.

Reply:
I don't think you need `Option` for elements of the tree. I think that there should be no empty nodes if you know the size of the tree?

Reply:
If we want to take advantage of the properties of complete trees, we need some kind of "empty node".
Another way is to use `Hash::zeroed().typed()`

Reply:
I don't understand why. Here is how I see it. Complete tree should:
* be backed with a vec of capacity `2^height`
* double the size every time `vec.len` is about to become `== vec.cap - 1`. Last element is always unused. Due to this, you'd have to be careful when calculating tree height out of `vec`'s capacity.
* number of nodes in the tree should always be `== vec.len` and all "empty" nodes are logically after `vec.len` and never accessed, i.e. in undefined state

Reply:
For example, could I represent the tree in my description differently than `[9b, 0a, 05, fc, 17, 05, **, 01, 02, 03, 04, 05]`?
I can't think of a way to achieve both having no empty nodes in the middle of the array and having the tree be complete (or having utility methods equivalent to `parent()`, `l_child()`, etc.)

Reply:
Are there empty nodes in the middle? I didn't think there should be any for complete tree. I'll take a closer look into implementation then

Reply:
The tree in your example isn't a complete tree

Reply:
Since a node is represented as `Option<HashOf<T>>` at the moment, it is a complete tree in my mind if I believe the following definitions:
https://en.wikipedia.org/wiki/Binary_tree#Types_of_binary_trees
https://towardsdatascience.com/5-types-of-binary-tree-with-cool-illustrations-9b335c430254

> consider using `zeroed` hash

Yes. We would explore tradeoffs between time and space efficiency (skipping hash calculations and not having wrappers)

Reply:
I'd create an issue for the remaining comments and merge the PR.

Reply:
#2396